### ISSUE TITLE
why my best.pt modelsize is 374.2M

### ISSUE THREAD

**Issue by ts19970816** [ultralytics/yolov5#4349]
parser.add_argument('--weights', type=str, default='./data/yolov5l.pt', help='initial weights path')![image](https://user-images.githubusercontent.com/51779812/128665980-7e9569a1-9934-4e37-ac26-ca5fff1d35c8.png)    parser.add_argument('--cfg', type=str, default='', help='model/yolov5l.yaml path')
    parser.add_argument('--data', type=str, default='data/voc.yaml', help='dataset.yaml path')
    parser.add_argument('--hyp', type=str, default='data/hyps/hyp.scratch.yaml', help='hyperparameters path')
    parser.add_argument('--epochs', type=int, default=300)
    parser.add_argument('--batch-size', type=int, default=1, help='total batch size for all GPUs')
    parser.add_argument('--imgsz', '--img', '--img-size', type=int, default=640, help='train, val image size (pixels)')
    parser.add_argument('--rect', action='store_true', help='rectangular training')
    parser.add_argument('--resume', nargs='?', const=True, default=False, help='resume most recent training')
    parser.add_argument('--nosave', action='store_true', help='only save final checkpoint')
    parser.add_argument('--noval', action='store_true', help='only validate final epoch')
    parser.add_argument('--noautoanchor', action='store_true', help='disable autoanchor check')
    parser.add_argument('--evolve', type=int, nargs='?', const=300, help='evolve hyperparameters for x generations')
    parser.add_argument('--bucket', type=str, default='', help='gsutil bucket')
    parser.add_argument('--cache', type=str, nargs='?', const='ram', help='--cache images in "ram" (default) or "disk"')
    parser.add_argument('--image-weights', action='store_true', help='use weighted image selection for training')
    parser.add_argument('--device', default='', help='cuda device, i.e. 0 or 0,1,2,3 or cpu')
    parser.add_argument('--multi-scale', action='store_true', help='vary img-size +/- 50%%')
    parser.add_argument('--single-cls', action='store_true', help='train multi-class data as single-class')
    parser.add_argument('--adam', action='store_true', help='use torch.optim.Adam() optimizer')
    parser.add_argument('--sync-bn', action='store_true', help='use SyncBatchNorm, only available in DDP mode')
    parser.add_argument('--workers', type=int, default=0, help='maximum number of dataloader workers')
    parser.add_argument('--project', default='runs/train', help='save to project/name')
    parser.add_argument('--entity', default=None, help='W&B entity')
    parser.add_argument('--name', default='exp', help='save to project/name')
    parser.add_argument('--exist-ok', action='store_true', help='existing project/name ok, do not increment')
    parser.add_argument('--quad', action='store_true', help='quad dataloader')
    parser.add_argument('--linear-lr', action='store_true', help='linear LR')
    parser.add_argument('--label-smoothing', type=float, default=0.0, help='Label smoothing epsilon')
    parser.add_argument('--upload_dataset', action='store_true', help='Upload dataset as W&B artifact table')
    parser.add_argument('--bbox_interval', type=int, default=-1, help='Set bounding-box image logging interval for W&B')
    parser.add_argument('--save_period', type=int, default=-1, help='Log model after every "save_period" epoch')
    parser.add_argument('--artifact_alias', type=str, default="latest", help='version of dataset artifact to be used')
    parser.add_argument('--local_rank', type=int, default=-1, help='DDP parameter, do not modify')
    parser.add_argument('--freeze', type=int, default=0, help='Number of layers to freeze. backbone=10, all=24')
    opt = parser.parse_known_args()[0] if known else parser.parse_args()
    return opt

![2021-08-09 14-09-40屏幕截图](https://user-images.githubusercontent.com/51779812/128666032-71f1dc26-918b-48d7-acdd-289123fe3736.png)

**Comment by iceisfun** [comment]
did you let training run to completion or interrupt it?

when it completes it will do a optimizer strip to reduce file size.

each epoch of training it will write a last.pt and possibly a best.pt which is much larger than the final result until the stripped output is written.
* mid training
$ du -hs debug/train/exp/weights/last.pt
55M	debug/train/exp/weights/last.pt

* after training
$ du -hs debug/train/exp/weights/last.pt
14M	debug/train/exp/weights/last.pt


```
5 epochs completed in 0.024 hours.

Optimizer stripped from runs/train/exp/weights/last.pt, 14.4MB
Optimizer stripped from runs/train/exp/weights/best.pt, 14.4MB
```

With 5x

667M	debug/train/exp/weights/last.pt
after stripping
167M	debug/train/exp/weights/last.pt

```
Optimizer stripped from runs/train/exp/weights/last.pt, 175.1MB
Optimizer stripped from runs/train/exp/weights/best.pt, 175.1MB
```

**Comment by ts19970816** [comment]
I didn't find  debug my best.pt or last.pt

**Comment by ts19970816** [comment]
What is debug

**Comment by mxy5201314** [comment]
> What is debug
Don't tease me

**Comment by ts19970816** [comment]
I don't know how to do it

**Comment by mxy5201314** [comment]
> 我不知道该怎么做

come on boy,  go to learn more

**Comment by ts19970816** [comment]
Big guy, teach me how to reduce the parameter size

**Comment by mxy5201314** [comment]
At the end of the training, the code will delete redundant information for you. Are you sure you have completed the training?

**Comment by ts19970816** [comment]
no I haven't completed a complete training

**Comment by mxy5201314** [comment]
If you train completely once, your parameters will be reduced

**Comment by ts19970816** [comment]
ok,thanks

**Comment by iceisfun** [comment]
Sorry, debug was just the output directory my training was saved in.

**Comment by ts19970816** [comment]
> Sorry, debug was just the output directory my training was saved in.

I do not understand you

**Comment by iceisfun** [comment]
debug has nothing to do with the model stripping process, it is just a folder.

to resolve your model size issue train until all epochs are complete.


```
$ ls -al
drwxr-xr-x  2 root root 4096 Aug 10 15:03 debug
$ 
```

**Comment by ts19970816** [comment]
ok,thanks

**Comment by glenn-jocher** [comment]
@ts19970816 YOLOv5 🚀 training checkpoints should be 4X the size of final trained checkpoints, as they carry not just a FP16 model, but a FP16 [EMA](https://www.tensorflow.org/api_docs/python/tf/train/ExponentialMovingAverage) and an FP32 optimizer of the same size as the model (each model parameter has its own FP32 gradient saved within the optimizer). Final checkpoints contain only an FP16 model, with the EMA and optimizer both stripped after the final epoch of training:
https://github.com/ultralytics/yolov5/blob/75feeb797c4a9553f4274860c6c278f1fc628f60/train.py#L435-L439

The `strip_optimizer()` function updates the checkpoint dictionary value for `model`, replacing it with `ema`, and sets `ema` and `optimizer` keys to `None`, which will reduce the checkpoint size by 3/4.
https://github.com/ultralytics/yolov5/blob/75feeb797c4a9553f4274860c6c278f1fc628f60/utils/general.py#L495-L509

Final trained checkpoints after `strip_optimizer()` should match the README table sizes below.
<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>
